i2c: Asynchronous Transfers

teburd opened the pull request:
Provides i2c_transfer_cb as an optional API along with a wrapper that provides i2c_transfer_signal (async with a k_poll_signal notifier).

Notably i2c_transfer_cb should not block so it can be initiated from an ISR if so desired as done in the frdm_k64f example. If it would block, returning -EWOULDBLOCK is expected.

aunsbjerg commented on subsys/async/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_ASYNC)
+  zephyr_sources(async.c)
```
Could use `zephyr_sources_ifdef(CONFIG_ASYNC async.c)` instead.

aunsbjerg commented on include/zephyr/async.h:
```diff
@@ -0,0 +1,99 @@
+ *
+ * @param result The result of the asynchronous call
+ * @param userdata Generic userdata requested to be tied
+ *        to the calback.
```
`calback` -> `callback`

aunsbjerg commented on include/zephyr/async.h:
```diff
@@ -0,0 +1,99 @@
+ * Semaphore completion object with return code.
+ */
+struct async_sem {
+	/** Semaphore that be given on completion */
```
Missing `will`

aunsbjerg commented on include/zephyr/drivers/i2c.h:
```diff
@@ -599,6 +607,63 @@ static inline int z_impl_i2c_transfer(const struct device *dev,
+ * to another I2C device asynchronously.
+ *
+ * The array of message @a msgs must not be NULL.  The number of
+ * message @a num_msgs may be zero,in which case no transfer occurs.
```
`zero,in` -> `zero, in`

teburd replied:
removed and replaced with @see to reference the doc string for i2c_transfer

aunsbjerg commented on include/zephyr/drivers/i2c.h:
```diff
@@ -599,6 +607,63 @@ static inline int z_impl_i2c_transfer(const struct device *dev,
+		return -ENOSYS;
+	}
+
+	int res =  api->transfer_async(dev, msgs, num_msgs, addr, cb, userdata);
```
Can return `api->transfer_async()` directly

aunsbjerg commented on include/zephyr/drivers/i2c.h:
```diff
@@ -599,6 +607,63 @@ static inline int z_impl_i2c_transfer(const struct device *dev,
+ * @param msgs Array of messages to transfer.
+ * @param num_msgs Number of messages to transfer.
+ * @param addr Address of the I2C target device.
+ * @param callback Function pointer for completion callback. See `async.h` for
```
`@param callback` -> `@param cb` to align with prototype and avoid doc build failure in CI

Commit:
i2c: Adds i2c_transfer_async API for I2C

i2c_transfer_async is the asynchronous version of i2c_transfer where
the completion of the transfer is notified to the caller using a callback.

This can be used in conjuction with the common callbacks and datatypes
in async.h for directly doing an async transfer with an IPC object
to notify a thread.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
i2c: mcux transfer with callbacks

Implements i2c_transfer_cb for the mcux driver

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
tests: boards: frdm_k64f i2c test against fxos8700

Adds a simple test case against the FXOS8700 sensor and its
built in FIFO using i2c. Pulls data from the sensor using both
synchronous and asynchronous transfers.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

tbursztyka commented on include/zephyr/async.h:
```diff
@@ -0,0 +1,107 @@
+/*
```
Could you split the API introduction to zephyr and the test into 2 commits?

tbursztyka commented on include/zephyr/async.h:
```diff
@@ -0,0 +1,107 @@
+/**
+ * Semaphore completion object with return code.
+ */
+struct async_sem {
```
Why not having 1 structure for all, and with a name like "async_notifier"?

Sure, k_poll_signal is not going to use the result attribute but it's only 4 bytes but it's not like there will be hundreds of such objects.

And then would be nice to have macros to set up each type of notifier (so the user does not have to go through the burden of manipulating sem/mutex/k_work/etc... by himself).
Perhaps also a macro or iniline as a helper to give the right generic callback for a selected type of notifier. 

Idea is that it could be made easy in user code to switch to another notifier.

teburd replied:
I'll consider what this means

tbursztyka commented on include/zephyr/async.h:
```diff
@@ -0,0 +1,107 @@
+ */
+struct async_work {
+	/** Work that will be submitted on completion */
+	struct k_work work;
```
why not a pointer like the other?

teburd replied:
For k_work to get the result, I used CONTAINER_OF. Probably could all be non-pointer with a macro for each to make it easy to define statically each one.

tbursztyka replied:
the user knows that the work is tied to this async struct (as for the other types, specifically the sem and mutex to get the result) so I don't see any real gain in user CONTAINER_OF here. 

I'd rather generalize on one struct made of pointers instead of specializing for each notifier type. Unless there is really a drawback in this approach I am missing

tbursztyka commented on include/zephyr/async.h:
```diff
@@ -0,0 +1,107 @@
+ * @param result Result to assign to the async_sem
+ * @param userdata Pointer to a struct async_sem
+ */
+void async_semaphore_fn(int result, void *userdata);
```
this is a callback: s/_fn/_cb 

Apply that everywhere

tbursztyka commented on tests/subsys/async/async_api/src/main.c:
```diff
@@ -0,0 +1,89 @@
+			 ztest_unit_test(test_async_mutex),
+			 ztest_unit_test(test_async_work)
+			 );
+	ztest_run_test_suite(test_async);
```
Test would benefit to have a dedicated thread to call the async callbacks. It's more work though, let's see

tbursztyka reviewed (commented):
Small comments on API side. i²c part seems fine (though perhaps generalizing the stuff like it was done in SPI would be beneficial, so having i2c_context.h etc..)

nordic-krch reviewed (commented):
Couple of comments

teburd commented on include/zephyr/async.h:
```diff
@@ -0,0 +1,107 @@
+ */
+typedef void (*async_callback_t)(int result, void *userdata);
+
+#ifdef CONFIG_ASYNC
```
It's not needed, but I've found the missing symbol linker error harder to grok than missing function prototype errors myself.

teburd commented on include/zephyr/async.h:
```diff
@@ -0,0 +1,107 @@
+/*
```
yeah perhaps, one of the issues with the proposal from @tbursztyka was the rom size increase because certain code paths couldn't be gc'ed as I read it.

I'll take a look at sys_notify. Really I wanted to avoid a single struct with a type tagged union for the same reasons. This really is mean to be a few helpers/types for dealing with callbacks and thread IPC mechanisms that are commonly used. async_xyz_cb and datatype are obvious and easy to add to imo.

nordic-krch replied:
if `async` approach is better i don't mind deprecating `notify`. I seems that it is not that widely used (only onoff now? was suppose to be used by my old i2c async propsal).

teburd commented:
split off the async helpers into #45562 and moved them to lib/os like sys_notify

Will revisit this once people look at it and are happy with it.

teburd commented:
In place of a common set of callback helpers, provide a direct k_poll_signal wrapper alongside the callback version of i2c_transfer, so adds i2c_transfer_cb (callback) and i2c_transfer_signal (k_poll_signal, wraps the callback API).

teburd commented:
I did spend some time trying to make the k_poll_signal i2c_transfer_signal a user space friendly syscall with some success. The downsides came to needing to alloc/free things when user space was being used to keep the life of a copy of the i2c_msg array around long enough.

It's certainly something that can be addressed in a future PR.

nordic-krch commented on include/zephyr/drivers/i2c.h:
```diff
@@ -179,6 +188,14 @@ typedef int (*i2c_api_target_register_t)(const struct device *dev,
 					  struct i2c_target_config *cfg);
+#ifdef CONFIG_I2C_CALLBACK
+typedef int (*i2c_api_transfer_cb_t)(const struct device *dev,
+				 struct i2c_msg *msgs,
```
i still think that it would be better to put all this into a struct. Struct can be later on extended without breaking api (e.g adding list node field if we go into direction of queued transfers).

Current version is suitable only for bus with one device since with more we might hit collision and handling collision is painful. At the end, it should be really fire and forget. Using struct gives path towards that without need of breaking the API in the future.

teburd replied:
This can be done by providing an asynchronous lock/unlock pair of functions

i2c_lock_async(dev, cb0, data);
i2c_transfer_cb(dev, i2c_msgs, cb1, data);
i2c_transfer_cb(dev, more_i2c_msgs, cb2, data);
i2c_unlock_async(dev, cb3, data);

Obviously not in a row like that but in a callback chain. Callback chaining like this is what RTIO makes look more like a queue of requests and pollable responses.

That's how I was planning on dealing with this issue rather than trying to embed a request queue directly in the device driver. In effect there's an application allocated request queue over all devices rather than each device having its own. This way you can do request chaining or parallelism with most of the I/O task concurrency that something like a promise pipeline might give you.
